README.md
# ffmpeg4android

To add FFmpeg support to your app:

1. Add this permission to your manifest: 
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>

2. Add this line to your app gradle.build:
implementation 'com.netcompss:ffmpeg4android_lib:41.06'

And you are done!

Test it by adding to your activity onCreate:

        GeneralUtils.checkForPermissionsMAndAbove(MainActivity.this, true);
        
        LoadJNI vk = new LoadJNI();
        
        try {
        
            String workFolder = getApplicationContext().getFilesDir().getAbsolutePath();
            
            String[] complexCommand = {"ffmpeg","-i", "/sdcard/videokit/in.mp4"};
            
            vk.run(complexCommand , workFolder , getApplicationContext());
            
            Log.i("test", "ffmpeg4android finished successfully");
            
        } catch (Throwable e) {
        
            Log.e("test", "vk run exception.", e);
            
        }
        
 
Make sure your device contains the in.mp4 file (small mp4 file for the tests), in this location:
/sdcard/videokit/in.mp4


release_notes.txt
4.1_08
-----
* Added android:extractNativeLibs="true" to the demo application manifest file

4.1_07
-----
* Fixed some bugs related to 32bit devices.

4.1_06
------
* removed x264 from command validation warnings.

4.1_05
------
* Upgraded base to 4.1.
* Loader improvements, much more reliable.
* Full 64-bit support.

3.22_03_ess_au_p27
--------------------
added adts muxer to support m4a encoding

-enable-parser=aac,dpx,pnm,aac_latm,dvaudio,ac3,flac,mpegaudio,vorbis,cook,dca,opus,dirac,dnxhd \
--enable-muxer=mp4,mp3,ogg,webm,wav,avi,adts \
--enable-demuxer=mp3,ogg,webm,wav,avi,mov,nut,aac,rawvideo \
--enable-decoder=mp3,aac,dvvideo,mjpeg,dnxhd,dirac,png,pcm_s16le \
--enable-encoder=png,jpeg2000,aac,libmp3lame,vorbis,pcm_s16le \

3.22.02
--------
* Progress  calculation fixes.
* Stability fixes.

3.22.01
--------
* Added Internet permission to support streaming (RealTime, and Manifest).

3.22.00
--------
* FFmpeg Base upgraded to 3.22
* Android Target updated to 25.

25.24.10 
------
* More cleanups

2.5.24.9
--------
* Refactors, and cleanups

2.5.24.8
--------
* checkForPermissionsMAndAbove block param added.

2.5.24.7
--------
* Added commands.txt, readme.txt, and more support text docs.
* checkForPermissionsMAndAboveBlocking added to support blocking permissions.


2.5.24.6
--------
* Fixed MultipleCommandsExample (missing png in the demo folder, and added -y).
* Fixed lib version number.
* CommandText fix for Android 6.

2.5.24.4
---------
* Removed x86 (moved to extras).
* abiFilter "x86" commented out

ffmpeg4android_lib/src/main/java/com/netcompss/loader/LoadJNI.java
package com.netcompss.loader;

import java.io.File;

import com.netcompss.ffmpeg4android.CommandValidationException;
import com.netcompss.ffmpeg4android.GeneralUtils;
import com.netcompss.ffmpeg4android.Prefs;

import android.content.Context;
import android.content.pm.PackageManager;
import android.util.Log;

public final class LoadJNI {

	static {
		System.loadLibrary("loader-jni");
	}
	
	/**
	 * 
	 * @param args ffmpeg command
	 * @param workFolder working directory 
	 * @param ctx Android context
	 * @param isValidate apply validation to the command
	 * @throws CommandValidationException
	 */
	public void run(String[] args, String workFolder, Context ctx, boolean isValidate) throws CommandValidationException {
		Log.i(Prefs.TAG, "running ffmpeg4android_lib: " + Prefs.version);
		// delete previous log: this is essential for correct progress calculation
		String vkLogPath = workFolder + "vk.log";
		GeneralUtils.deleteFileUtil(vkLogPath);
		GeneralUtils.printCommand(args);
		
		//printInternalDirStructure(ctx);
		
		if (isValidate) {
			if (GeneralUtils.isValidCommand(args))
				load(args, workFolder, getVideokitLibPath(ctx), true);
			else
				throw new CommandValidationException();
		}
		else {
			load(args, workFolder, getVideokitLibPath(ctx), true);
		}
		
	}
	
	/**
	 * 
	 * @param args ffmpeg command
	 * @param workFolder working directory
	 * @param ctx Android context
	 * @throws CommandValidationException
	 */
	public void run(String[] args, String workFolder, Context ctx) throws CommandValidationException {
		run(args, workFolder, ctx, true);
	}
	
	
	private static void printInternalDirStructure(Context ctx) {
		Log.d(Prefs.TAG, "=printInternalDirStructure=");
		Log.d(Prefs.TAG, "==============================");
		File file = new File(ctx.getFilesDir().getParent());
		analyzeDir(file);
		Log.d(Prefs.TAG, "==============================");
	}
	
	private static void analyzeDir(File path) {
		if (path.isDirectory()) {
			Log.d(Prefs.TAG, "Scanning dir: " + path.getAbsolutePath());
			File[] files1 = path.listFiles();
			for (int i = 0; i < files1.length; i++) {
				analyzeDir(files1[i]);
			}
			Log.d(Prefs.TAG, "==========");
		}
		else {
			Log.w(Prefs.TAG, path.getAbsolutePath() + " not a dir.");

		}
	}
	
	private static String getVideokitLibPath(Context ctx) {

		// working 64 bit
		String videokitLibPath = ctx.getApplicationInfo().nativeLibraryDir  + "/libvideokit.so";

		File file = new File(videokitLibPath);
		if(file.exists())  {     
		    Log.i(Prefs.TAG, "videokitLibPath exits");
			Log.i(Prefs.TAG, videokitLibPath);
		}
		else {
			Log.e(Prefs.TAG, "videokitLibPath not exits: " + videokitLibPath);

		}

		return videokitLibPath;
		
	}
	
	
	
	public void fExit( Context ctx) {
		fexit(getVideokitLibPath(ctx));
	}


	@SuppressWarnings("JniMissingFunction")
	public native String fexit(String videokitLibPath);

	@SuppressWarnings("JniMissingFunction")
	public native String unload();

	@SuppressWarnings("JniMissingFunction")
	public native String load(String[] args, String videokitSdcardPath, String videokitLibPath, boolean isComplex);

}


ffmpeg4android_demo/src/main/java/com/examples/ffmpeg4android/Main.java
package com.examples.ffmpeg4android;

import com.examples.ffmpeg4android_demo.R;
import com.netcompss.ffmpeg4android.GeneralUtils;
import com.netcompss.ffmpeg4android.Prefs;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;

public class Main extends Activity  {
	
	@Override
	protected void onResume() {
		super.onResume();
		Log.i(Prefs.TAG, "Main on resume handling log copy in case of a crash");
		String demoVideoFolder = Environment.getExternalStorageDirectory().getAbsolutePath() + "/videokit/";
		String workFolder = getApplicationContext().getFilesDir() + "/";
		String vkLogPath = workFolder + "vk.log";
		GeneralUtils.copyFileToFolder(vkLogPath, demoVideoFolder);

	}
	
	@Override
	  public void onCreate(Bundle savedInstanceState) {
	      super.onCreate(savedInstanceState);
	      setContentView(R.layout.main);
	      GeneralUtils.checkForPermissionsMAndAbove(Main.this, false);
	      
	      Button simpleAct =  (Button)findViewById(R.id.startSimpleAct);
	      simpleAct.setOnClickListener(new OnClickListener() {
				public void onClick(View v){
					Log.i(Prefs.TAG, "run simpleAct.");
					startAct(SimpleExample.class);
					
				}
		  });
	      
	      Button progressAct =  (Button)findViewById(R.id.startProgressAct);
	      progressAct.setOnClickListener(new OnClickListener() {
				public void onClick(View v){
					Log.i(Prefs.TAG, "run progressAct.");
					startAct(ProgressBarExample.class);
					
				}
		  });
	      
	      Button progressWithNotificationAct =  (Button)findViewById(R.id.startProgressWithNotificationAct);
	      progressWithNotificationAct.setOnClickListener(new OnClickListener() {
				public void onClick(View v){
					Log.i(Prefs.TAG, "run progressWithNotificationAct.");
					startAct(ProgressBarWithNotificationExample.class);
				}
		  });
	      
	      Button multiAct =  (Button)findViewById(R.id.startMultiAct);
	      multiAct.setOnClickListener(new OnClickListener() {
				public void onClick(View v){
					Log.i(Prefs.TAG, "run MultiplyCommandsExample.");
					startAct(MultipleCommandsExample.class);
				}
		  });
	      
	      
	}
	
	public static String getAboutText(Context ctx) {
    	StringBuffer b = new StringBuffer();
    	b.append("\n" + ctx.getString(R.string.full_app_name) + "\n");
    	b.append("Developed by: NetComps LTD\n" );
    	b.append("Version: " + GeneralUtils.getVersionName(ctx) + "\n");
    	b.append("Support: " + ctx.getString(R.string.email) + "\n");
    	return b.toString();
    }
	
	
	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.main, menu);
		return true;
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		case R.id.action_about:
    		AlertDialog.Builder builder = new AlertDialog.Builder(this);
    		builder.setTitle("About").
    		setMessage(getAboutText(getApplicationContext())).
    		setPositiveButton("OK", new DialogInterface.OnClickListener() {
    	           public void onClick(DialogInterface dialog, int id) {
    	               // User clicked OK button
    	           }
    	       });
    		
    		AlertDialog about  = builder.create();
    		about.show();
     		return true;
		default:
			return super.onOptionsItemSelected(item);
		}
	}
	
	private void startAct(Class act) {
		Intent intent = new Intent(this, act);
		Log.d(Prefs.TAG, "Starting act:" + act);
		this.startActivity(intent);
	}
	
	


}


ffmpeg4android_demo/src/main/java/com/examples/ffmpeg4android/SimpleExample.java
package com.examples.ffmpeg4android;


import com.examples.ffmpeg4android_demo.R;
import com.netcompss.ffmpeg4android.CommandValidationException;
import com.netcompss.ffmpeg4android.GeneralUtils;
import com.netcompss.ffmpeg4android.Prefs;
import com.netcompss.loader.LoadJNI;
import android.app.Activity;
import android.app.ProgressDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.os.PowerManager;
import android.os.PowerManager.WakeLock;

import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

/**
 *  To run this Demo Make sure you have on your device this folder:
 *  /sdcard/videokit,
 *  and you have in this folder a video file called in.mp4
 * @author elih
 *
 */
public class SimpleExample extends Activity {

	String workFolder = null;
	String demoVideoFolder = null;
	String demoVideoPath = null;
	String vkLogPath = null;
	private boolean commandValidationFailedFlag = false;


	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.ffmpeg_demo_client_1);

		demoVideoFolder = Environment.getExternalStorageDirectory().getAbsolutePath() + "/videokit/";
		demoVideoPath = demoVideoFolder + "in.mp4";

		Log.i(Prefs.TAG, getString(R.string.app_name) + " version: " + GeneralUtils.getVersionName(getApplicationContext()) );
		workFolder = getApplicationContext().getFilesDir().getAbsolutePath() + "/";
		//Log.i(Prefs.TAG, "workFolder: " + workFolder);
		vkLogPath = workFolder + "vk.log";

		GeneralUtils.copyLicenseFromAssetsToSDIfNeeded(this, workFolder);
		GeneralUtils.copyDemoVideoFromAssetsToSDIfNeeded(this, demoVideoFolder);

		Button invoke =  (Button)findViewById(R.id.invokeButton);
		invoke.setOnClickListener(new OnClickListener() {
			public void onClick(View v){
				Log.i(Prefs.TAG, "run clicked.");
				if (GeneralUtils.checkIfFileExistAndNotEmpty(demoVideoPath)) {
					new TranscdingBackground(SimpleExample.this).execute();
				}
				else {
					Toast.makeText(getApplicationContext(), demoVideoPath + " not found", Toast.LENGTH_LONG).show();
				}
			}
		});

		int rc = GeneralUtils.isLicenseValid(getApplicationContext(), workFolder);
		Log.i(Prefs.TAG, "License check RC: " + rc);
	}

	public class TranscdingBackground extends AsyncTask<String, Integer, Integer>
	{

		ProgressDialog progressDialog;
		Activity _act;
		String commandStr;

		public TranscdingBackground (Activity act) {
			_act = act;
		}



		@Override
		protected void onPreExecute() {
			EditText commandText = (EditText)findViewById(R.id.CommandText);
			commandStr = commandText.getText().toString();

			progressDialog = new ProgressDialog(_act);
			progressDialog.setMessage("FFmpeg4Android Transcoding in progress...");
			progressDialog.show();

		}

		protected Integer doInBackground(String... paths) {
			Log.i(Prefs.TAG, "doInBackground started...");

			// delete previous log
			boolean isDeleted = GeneralUtils.deleteFileUtil(workFolder + "/vk.log");
			Log.i(Prefs.TAG, "vk deleted: " + isDeleted);

			PowerManager powerManager = (PowerManager)_act.getSystemService(Activity.POWER_SERVICE);
			WakeLock wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "VK_LOCK");
			Log.d(Prefs.TAG, "Acquire wake lock");
			wakeLock.acquire();

			///////////// Set Command using code (overriding the UI EditText) /////
			//commandStr = "ffmpeg -y -i /sdcard/videokit/in.mp4 -strict experimental -s 320x240 -r 30 -aspect 3:4 -ab 48000 -ac 2 -ar 22050 -vcodec mpeg4 -b 2097152 /sdcard/videokit/out.mp4";
			//String[] complexCommand = {"ffmpeg", "-y" ,"-i", "/sdcard/videokit/in.mp4","-strict","experimental","-s", "160x120","-r","25", "-vcodec", "mpeg4", "-b", "150k", "-ab","48000", "-ac", "2", "-ar", "22050", "/sdcard/videokit/out.mp4"};
			///////////////////////////////////////////////////////////////////////


			LoadJNI vk = new LoadJNI();
			try {

				// complex command
				//vk.run(complexCommand, workFolder, getApplicationContext());

				vk.run(GeneralUtils.utilConvertToComplex(commandStr), workFolder, getApplicationContext());

				// running without command validation
				//vk.run(complexCommand, workFolder, getApplicationContext(), false);

				// copying vk.log (internal native log) to the videokit folder
				GeneralUtils.copyFileToFolder(vkLogPath, demoVideoFolder);

			} catch (CommandValidationException e) {
					Log.e(Prefs.TAG, "vk run exeption.", e);
					commandValidationFailedFlag = true;

			} catch (Throwable e) {
				Log.e(Prefs.TAG, "vk run exeption.", e);
			}
			finally {
				if (wakeLock.isHeld())
					wakeLock.release();
				else{
					Log.i(Prefs.TAG, "Wake lock is already released, doing nothing");
				}
			}
			Log.i(Prefs.TAG, "doInBackground finished");
			return Integer.valueOf(0);
		}

		protected void onProgressUpdate(Integer... progress) {
		}

		@Override
		protected void onCancelled() {
			Log.i(Prefs.TAG, "onCancelled");
			//progressDialog.dismiss();
			super.onCancelled();
		}


		@Override
		protected void onPostExecute(Integer result) {
			Log.i(Prefs.TAG, "onPostExecute");
			progressDialog.dismiss();
			super.onPostExecute(result);

			// finished Toast
			String rc = null;
			if (commandValidationFailedFlag) {
				rc = "Command Vaidation Failed";
			}
			else {
				rc = GeneralUtils.getReturnCodeFromLog(vkLogPath);
			}
			final String status = rc;
			SimpleExample.this.runOnUiThread(new Runnable() {
				public void run() {
					Toast.makeText(SimpleExample.this, status, Toast.LENGTH_LONG).show();
					if (status.equals("Transcoding Status: Failed")) {
						Toast.makeText(SimpleExample.this, "Check: " + vkLogPath + " for more information.", Toast.LENGTH_LONG).show();
					}
				}
			});
		}

	}


}


ffmpeg4android_lib/src/main/java/com/netcompss/ffmpeg4android/GeneralUtils.java
package com.netcompss.ffmpeg4android;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.util.ArrayList;

//import com.examples.ffmpeg4android_demo_native.R;

import android.Manifest;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Build;
import android.util.Log;
import android.widget.Toast;


public class GeneralUtils {
	
	public static String[] fixComplexCommand(String[] complexCommand) {
		for (int i = 0; i < complexCommand.length; i++) {
			if (complexCommand[i].startsWith("\"") || complexCommand[i].startsWith(" ")) {
				StringBuffer sb = new StringBuffer();
				sb.append(complexCommand[i]);
				
				while (sb.charAt(0) == ' ' || sb.charAt(0) == '"' ) {
					sb.deleteCharAt(0);
				}
				while (sb.charAt(sb.length() - 1) == ' ' || sb.charAt(sb.length() - 1) == '"') {
					sb.deleteCharAt(sb.length() - 1);
				}
				
				complexCommand[i] =  sb.toString();
				Log.d(Prefs.TAG, "command " + i + ": " + sb.toString());
			}
		}
		return complexCommand;
	}
	
	public static void printCommand(String[] command) {
		StringBuffer sb = new StringBuffer();
		sb.append("{");
		for (int i = 0; i < command.length; i++ ) {
			sb.append("\"");
			sb.append(command[i]);
			sb.append("\"");
			if (i < command.length -1)
			  sb.append(",");
		}
		sb.append("}");
		Log.d(Prefs.TAG, sb.toString());
	}
	
	
	// {"ffmpeg", "-y" ,"-i", "/sdcard/videokit/sample.mp4",
	//  "-strict","experimental","-s", "160x120","-r","25", "-vcodec", "mpeg4",
	//  "-b", "150k", "-ab","48000", "-ac", "2", "-ar", "22050", "/sdcard/videokit/out.mp4"};
	public static boolean isValidCommand(String[] args) {
		/*
		for (int i = 0; i < args.length; i++) {
			if (args[i].equals("libx264") || args[i].equals("-preset")) {
				Log.w(Prefs.TAG, "Command validation detected libx264 use");
				Log.w(Prefs.TAG, "Make sure you use the extra libs that support libx264");
				//Log.w(Prefs.TAG, "Note that you will have to target api 22 or below in your Manifest when using libx264");
				break;
			}
		}
		*/
		
		// find the input files path
		ArrayList<Integer> inputFilesArrLocation = new ArrayList<Integer>();
		for (int i = 0; i < args.length; i++) {
			if (args[i].equals("-i"))
				inputFilesArrLocation.add(i+1);
		}
		
		// verify the input files exists
		boolean okFlag = true;
		for (int i = 0; i < inputFilesArrLocation.size(); i++) {
			String path = args[inputFilesArrLocation.get(i)];
			if (!(okFlag = checkIfFileExistAndNotEmpty(path))) {
				Log.e(Prefs.TAG, "Command validation failed.");
				Log.e(Prefs.TAG, "Check if input file exists: " + path);
				return okFlag;
			}
		}
		
		
		String outputPath = args[args.length -1];
		if (isStream(outputPath)){
			Log.i(Prefs.TAG, "output is a stream");
			return okFlag;
		}
		
		// verify output folder exists
		int indexOfLastSlash = outputPath.lastIndexOf("/");
		if (indexOfLastSlash == -1) {
			Log.e(Prefs.TAG, "Command validation failed.");
			Log.e(Prefs.TAG, "No slashes in output path looks like the : " + outputPath + " is not valid.");
			return false;
		}
		String outputFolder = outputPath.substring(0, indexOfLastSlash);
		if (!(okFlag = checkIfFolderExists(outputFolder))) {
			Log.e(Prefs.TAG, "Command validation failed.");
			Log.e(Prefs.TAG, "Check if output folder exists: " + outputFolder);
		}
		
		return okFlag;
	}
	
	public static boolean isStream(String filePath) {
		// handling streaming case e.g: udp://
		if (filePath.startsWith("udp://")) {
			Log.i(Prefs.TAG,"mached stream");
			return true;
		}
		else return false;
	}
	
	@TargetApi(23) 
	public static void checkForPermissionsMAndAbove(Activity act, boolean isBlocking) {
		Log.i(Prefs.TAG, "checkForPermissions() called");
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
			// Here, thisActivity is the current activity
			if (	act.checkSelfPermission(
						Manifest.permission.WRITE_EXTERNAL_STORAGE)
						!= PackageManager.PERMISSION_GRANTED ||

					act.checkSelfPermission(
							Manifest.permission.INTERNET)
							!= PackageManager.PERMISSION_GRANTED
				)

			{


				// No explanation needed, we can request the permission.
				act.requestPermissions(
						new String[]{
								Manifest.permission.WRITE_EXTERNAL_STORAGE,
								Manifest.permission.INTERNET
						},
						0);
				if (isBlocking) {
					while (true) {
						if (act.checkSelfPermission(
								Manifest.permission.WRITE_EXTERNAL_STORAGE)
								== PackageManager.PERMISSION_GRANTED) {

							Log.i(Prefs.TAG, "Got permissions, exiting block loop");
							break;
						}
						Log.i(Prefs.TAG, "Sleeping, waiting for permissions");
						try {
							Thread.sleep(1000);
						} catch (InterruptedException e) {
							e.printStackTrace();
						}
					}
				}

			}
			// permission already granted
			else {
				Log.i(Prefs.TAG, "permission already granted");
			}
		}
		else {
			Log.i(Prefs.TAG, "Below M, permissions not via code");
		}

    }
    
	
	public static long getVKLogSizeRandomAccess(String vkLogpath) {
		RandomAccessFile f = null;
		long ret = -1;
		try {
			f = new RandomAccessFile(vkLogpath, "r");
			ret = f.length();
			f.close();
		} catch (FileNotFoundException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());

		} catch (IOException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());
		}
		
		return ret;
	}
	
	@SuppressLint("DefaultLocale") 
	public static boolean isValidPicExtension(String fileName) {
		int endIndex = fileName.lastIndexOf(".");
		String ext = fileName.substring(endIndex + 1);
		ext = ext.toLowerCase();
		if (    (ext.toLowerCase().equals("jpg") ) ||
				(ext.toLowerCase().equals("bmp") ) ||
				(ext.toLowerCase().equals("png") ) ||
				(ext.toLowerCase().equals("jpeg") ) 
				) {
			return true;
		}
		else {
			return false;
		}

	}

	public static boolean checkIfFileExistAndNotEmpty(String fullFileName) {
		// the case of input images with %d: /sdcard/videokit/pic%03d.jpg
		if (fullFileName.contains("%") && isValidPicExtension(fullFileName)) {
			Log.i(Prefs.TAG,"mached picture array input");
			return true;
		}
	
		if (isStream(fullFileName)) return true;
	
		///////////////////////////////////
		File f = new File(fullFileName);
		long lengthInBytes = f.length();
		Log.d(Prefs.TAG, fullFileName + " length in bytes: " + lengthInBytes);
		if (lengthInBytes > 100)
			return true;
		else {
			return false;
		}

	}
	
	public static boolean deleteFile(String fullFileName) {
		File f = new File(fullFileName);
		boolean isdeleted = f.delete();
		return isdeleted;
	}

	public static boolean deleteFileUtil(String fullFileName) {
		File f = new File(fullFileName);
		boolean isdeleted = f.delete();
		return isdeleted;
		//Log.d(Prefs.TAG, "deleteing: " + fullFileName + " isdeleted: " + isdeleted);
	}

	public static String getReturnCodeFromLog(String filePath)  {
		String status = "Transcoding Status: Unknown";

		RandomAccessFile f = null;
		try {
			f = new RandomAccessFile(filePath, "r");
			long endLocation  = f.length();
			long seekLocation = -1;
			if (( seekLocation = endLocation - 100) < 0) {
				seekLocation = 0;
			}
			f.seek(seekLocation);

			String line;
			while ((line = f.readLine()) != null){ 
				if (line.startsWith("ffmpeg4android: 0")) {
					status = "Transcoding Status: Finished OK";
					break;
				}
				else if (line.startsWith("ffmpeg4android: 1") ) {
					status = "Transcoding Status: Failed";
					break;
				}
				
				else if (line.startsWith("ffmpeg4android: 2") ) {
					status = "Transcoding Status: Stopped";
					break;
				}

			}
			f.close();
		} catch (Exception e) {
			Log.e(Prefs.TAG, e.getMessage());
		}

		return status;
	}
	
	public static String[] utilConvertToComplex(String str) {
		 String[] complex = str.split(" ");
		 return complex;
	 }
	
	public static String getDutationFromVCLogRandomAccess(String vkLogpath) {
		String duration = null;
		try {
			RandomAccessFile f = new RandomAccessFile(vkLogpath, "r");
			String line;
			//f.seek(0);
			
			while ((line = f.readLine()) != null){ 
				//Log.d(Prefs.TAG, line);
				// Duration: 00:00:04.20, start: 0.000000, bitrate: 1601 kb/s 
				int i1 = line.indexOf("Duration:");
				int i2 = line.indexOf(", start");
				if (i1 != -1 && i2 != -1) {
					duration = line.substring(i1 + 10, i2 );
					break;
				}
			}
			f.close();
		} catch (FileNotFoundException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());

		} catch (IOException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());
		}
		return duration;
	}
	
	public static String readLastTimeFromVKLogUsingRandomAccess(String vkLogPath) {
		String timeStr = "00:00:00.00";
		try {

			RandomAccessFile f = null;
			f = new RandomAccessFile(vkLogPath, "r");
			String line;
			long endLocation  = f.length();
			long seekLocation = -1;
			if (( seekLocation = endLocation - 100) < 0) {
				seekLocation = 0;
			}
			f.seek(seekLocation);
			//Log.d(Prefs.TAG, "Starting while loop seekLocation: " + seekLocation);

			while ((line = f.readLine()) != null){ 
				Log.i("line", line);
				int i1 = line.indexOf("time=");
				int i2 = line.indexOf("bitrate=");
				if (i1 != -1 && i2 != -1) {
					timeStr = line.substring(i1 + 5, i2 - 1);
					Log.d(Prefs.TAG, "Found timeStr: " + timeStr);
				}
				else if (line.startsWith("ffmpeg4android: 0")) {
					timeStr = "exit";
				}
				else if (line.startsWith("ffmpeg4android: 1") ) {
						Log.w(Prefs.TAG, "error line: " + line);
						Log.w(Prefs.TAG, "Looks like error in the log");
						timeStr = "error";
				}
			}
			f.close();
		} catch (FileNotFoundException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());

		} catch (IOException e) {
			Log.i(Prefs.TAG,"waiting for file to be created: " + e.getMessage());
		}
		return timeStr.trim();
		
	}
	
	public static void copyLicenseFromAssetsToSDIfNeeded(Activity act, String destinationFolderPath) {
		InputStream is = null;
		BufferedOutputStream o = null;
		boolean copyLic = true;
		File destLic = null;
		try {
			is = act.getApplication().getAssets().open("ffmpeglicense.lic");
		} catch (Exception e) {
			Log.i(Prefs.TAG, "License file does not exist in the assets.");
			copyLic = false;
		}

		if (copyLic) {
			destLic = new File(destinationFolderPath + "ffmpeglicense.lic");
			//Log.i(Prefs.TAG, "Adding lic file at " + destLic.getAbsolutePath());

			o = null;
			try {
				byte[] buff = new byte[10000];
				int read = -1;
				o = new BufferedOutputStream(new FileOutputStream(destLic), 10000);
				while ((read = is.read(buff)) > -1) { 
					o.write(buff, 0, read);
				}
				//Log.i(Prefs.TAG, "Copy " + destLic.getAbsolutePath() + " from assets to SDCARD finished succesfully");
			}
			catch (Exception e) {
				Log.e(Prefs.TAG, "Error when coping license file from assets to working folder: " + e.getMessage());
			}
			finally {
				try {
					is.close();
					if (o != null) o.close();
				} catch (IOException e) {
					Log.w(Prefs.TAG, "Error when closing license file io: " + e.getMessage());
				}  

			}

		}
		else {
			Log.i(Prefs.TAG, "Not coping license");
		}
	
	}
	
	public static String getValidFileNameFromPath(String path) {
		int startIndex = path.lastIndexOf("/") + 1;
		int endIndex = path.lastIndexOf(".");
		
		String name = path.substring(startIndex, endIndex);
		String ext = path.substring(endIndex + 1);
		Log.d(Prefs.TAG, "name: " + name + " ext: " + ext);
		String validName = (name.replaceAll("\\Q.\\E", "_")).replaceAll(" ", "_");
		return validName + "." + ext;
	}
	
	public static String copyFileToFolder(String filePath, String folderPath) {
		//Log.i(Prefs.TAG, "Coping file: " + filePath + " to: " + folderPath);
		String validFilePathStr = filePath;
		try {
			FileInputStream is = new FileInputStream(filePath); 
			BufferedOutputStream o = null;
			String validFileName = getValidFileNameFromPath(filePath);
			validFilePathStr = folderPath + validFileName;
			File destFile = new File(validFilePathStr);
			try {
				byte[] buff = new byte[10000];
				int read = -1;
				o = new BufferedOutputStream(new FileOutputStream(destFile), 10000);
				while ((read = is.read(buff)) > -1) { 
					o.write(buff, 0, read);
				}
			} finally {
				is.close();
				if (o != null) o.close();  

			}
		} catch (FileNotFoundException e) {
			Log.w(Prefs.TAG, e.getMessage());
		} catch (IOException e) {
			Log.w(Prefs.TAG, e.getMessage());
		}
		return validFilePathStr;
	}
	
	public static boolean checkIfFolderExists(String fullFileName) {
		File f = new File(fullFileName);
		//Log.d(Prefs.TAG,"Checking if : " +  fullFileName + " exists" );
		if (f.exists() && f.isDirectory()) {
			//Log.d(Prefs.TAG,"Direcory: " +  fullFileName + " exists" );
			return true;
		}
		else {
			return false;
		}
	}
	
	public static boolean createFolder(String folderPath) {
		File f = new File(folderPath);
		return f.mkdirs();
	}
	
	public static void copyDemoVideoFromAssetsToSDIfNeeded(Activity act, String destinationFolderPath) {
		File destVid = null;
		try {
			if (!GeneralUtils.checkIfFolderExists(destinationFolderPath)) {

				boolean isFolderCreated = GeneralUtils.createFolder(destinationFolderPath);
				Log.i(Prefs.TAG, destinationFolderPath + " created? " + isFolderCreated);
				if (isFolderCreated) {
					destVid = new File(destinationFolderPath + "in.mp4");
					Log.i(Prefs.TAG, "Adding vid file at " + destVid.getAbsolutePath());
					InputStream is = act.getAssets().open("in.mp4");
					BufferedOutputStream o = null;
					try {
						byte[] buff = new byte[10000];
						int read = -1;
						o = new BufferedOutputStream(new FileOutputStream(destVid), 10000);
						while ((read = is.read(buff)) > -1) { 
							o.write(buff, 0, read);
						}
						Log.i(Prefs.TAG, "Copy " + destVid.getAbsolutePath() + " from assets to SDCARD finished succesfully");
					} 
					catch (Exception e) {
						Log.w(Prefs.TAG, "Failed copying: " + destVid.getAbsolutePath());
					}
					finally {
						is.close();
						if (o != null) o.close(); 

					}

				}
				else {
					Log.w(Prefs.TAG, "Demo videos folder was not created.");
				}

			}
			else {
				Log.d(Prefs.TAG, "demo videos directory exists, not copying demo video)");

			}

		} catch (FileNotFoundException e) {
			Log.e(Prefs.TAG, e.getMessage());
		} catch (IOException e) {
			Log.e(Prefs.TAG, e.getMessage());
		}
	}
	
	public static String getVersionName(Context ctx) {
		/*
		String versionName = "";
		try {
			versionName = ctx.getPackageManager().getPackageInfo(ctx.getPackageName(), 0).versionName;
		} catch (NameNotFoundException e) {
			Log.w(Prefs.TAG, "No version code found, returning -1");
		}
		
		return versionName;
		*/
		return Prefs.version;
	}
	
	
	
	public static int isLicenseValid(Context ctx, String workingFolder) {
		 LicenseCheckJNI lm = new LicenseCheckJNI();
		  int rc = lm.licenseCheck(workingFolder, ctx);
		  if (rc >= 0) {
			  if (rc == 1) {
				  //Toast.makeText(ctx, "Permanent license.", Toast.LENGTH_SHORT).show();
			  }
			  else if (rc == 2 || rc == 0) {
				  //Toast.makeText(ctx, "Trial license.", Toast.LENGTH_SHORT).show();
			  }
		  }
		  else if (rc < 0) {
			  if (rc == -1)
				  Toast.makeText(ctx, "Trail Expired. contact support.", Toast.LENGTH_LONG).show();
			  else if (rc == -2) 
				  Toast.makeText(ctx, "License invalid contact support", Toast.LENGTH_LONG).show();
			  else 
				  Toast.makeText(ctx, "License check failed. contact support." + rc, Toast.LENGTH_LONG).show();
		  }
		  return rc;
	}


}
